src/main/java/io/spokestack/spokestack/nlu/NLUService.java

```diff
@@ -0,0 +1,52 @@
+package io.spokestack.spokestack.nlu;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A base class for components that provide intent classification and slot
+ * recognition, either on-device or via a network request.
+ *
+ * <p>
+ * Intent classification is designed to be an asynchronous process; in order
+ * to receive results as they are available, add a listener to the NLU
+ * service.
+ * </p>
+ */
+public abstract class NLUService {
+
+    /**
+     * Listeners that receive results from this component.
+     */
+    private List<NLUListener> listeners = new ArrayList<>();
+
+    /**
+     * Add an NLU listener to receive results from this component.
+     *
+     * @param listener The listener to add.
+     */
+    public void addListener(NLUListener listener) {
+        this.listeners.add(listener);
+    }
+
+    /**
+     * Classifies a user utterance into an intent.
+     *
+     * Implementations should ensure that {@link #dispatch(NLUResult)} is
+     * called when a result is available so that listeners are notified.
+     *
+     * @param request The user request to be classified.
+     */
+    public abstract void classify(NLURequest request);
```

Comment:
while it is good to have an asynchronous interface for intent classification when you have a client that doesn't want to wait, it would also be good to have a simple way to call the intent classifier and get the result immediately without having to use callbacks